waku/v2/protocol/waku_filter/protocol.nim

```diff
@@ -0,0 +1,282 @@
+
+  WakuFilterResult*[T] = Result[T, string]  
+
+  WakuFilter* = ref object of LPProtocol
+    rng*: ref BrHmacDrbgContext
+    peerManager*: PeerManager
+    pushHandler*: MessagePushHandler
+    subscriptions*: seq[Subscription]
+    failedPeers*: Table[string, chronos.Moment]
+    timeout*: chronos.Duration
+
+proc init(wf: WakuFilter) =
+
+  proc handle(conn: Connection, proto: string) {.async, gcsafe, closure.} =
+    let message = await conn.readLp(MaxRpcSize.int)
+
+    let res = FilterRPC.init(message)
+    if res.isErr():
+      waku_filter_errors.inc(labelValues = [decodeRpcFailure])
+      return
+    
+    let rpc = res.get()
+
+    ## Filter request
+    # We are receiving a subscription/unsubscription request
+    if rpc.request != FilterRequest():
+      waku_filter_messages.inc(labelValues = ["FilterRequest"])
+
+      let 
+        requestId = rpc.requestId
+        subscribe = rpc.request.subscribe
+        pubsubTopic =  rpc.request.pubsubTopic
+        contentTopics = rpc.request.contentFilters.mapIt(it.contentTopic)
+
+      if subscribe:
+        wf.subscriptions.addSubscription(conn.peerId, requestId, pubsubTopic, contentTopics)
+      else:
+        wf.subscriptions.removeSubscription(conn.peerId, contentTopics)
+
+      waku_filter_subscribers.set(wf.subscriptions.len.int64)
+      
+    
+    ## Push message
+    # We are receiving a messages from the peer that we subscribed to
+    if rpc.push != MessagePush():
+      waku_filter_messages.inc(labelValues = ["MessagePush"])
+      
+      let
+        requestId = rpc.requestId 
+        push = rpc.push
+
+      await wf.pushHandler(requestId, push)
+
+
+  wf.handler = handle
+  wf.codec = WakuFilterCodec
+
+proc init*(T: type WakuFilter, 
+           peerManager: PeerManager, 
+           rng: ref BrHmacDrbgContext, 
+           handler: MessagePushHandler,
+           timeout: Duration = WakuFilterTimeout): T =
+  let wf = WakuFilter(rng: rng,
+                      peerManager: peerManager, 
+                      pushHandler: handler,
+                      timeout: timeout)
+  wf.init()
+  return wf
+
+
+proc setPeer*(wf: WakuFilter, peer: RemotePeerInfo) =
+  wf.peerManager.addPeer(peer, WakuFilterCodec)
+  waku_filter_peers.inc()
+  
+
+proc sendFilterRpcToPeer(wf: WakuFilter, rpc: FilterRPC, peer: PeerId): Future[WakuFilterResult[void]] {.async, gcsafe.}=
+  let connOpt = await wf.peerManager.dialPeer(peer, WakuFilterCodec)
+  if connOpt.isNone():
+    return err(dialFailure)
+
+  let connection = connOpt.get()
+
+  await connection.writeLP(rpc.encode().buffer)
+
+  return ok()
+
+proc sendFilterRpcToRemotePeer(wf: WakuFilter, rpc: FilterRPC, peer: RemotePeerInfo): Future[WakuFilterResult[void]] {.async, gcsafe.}=
+  let connOpt = await wf.peerManager.dialPeer(peer, WakuFilterCodec)
+  if connOpt.isNone():
+    return err(dialFailure)
+
+  let connection = connOpt.get()
+
+  await connection.writeLP(rpc.encode().buffer)
+
+  return ok()

```

Comment:
Couldn't we get away with removing this entire proc and only using `sendFilterRpcToPeer()` with `peerId`? PeerID should be available from wherever we want to dial and we're only dialing existing peers. Similar comments under `subscribe` and `unsubscribe` below.

Reply:
I am glad that you both noticed. I also saw it, the refactoring and cleanup work has these good things. You notice that there are things that can be simplified and optimized even more. 

It took me like 1 day to notice that there are two methods called `dialPeer()` in the `peer_manager.nim` file:

* With `RemotePeerInfo` as an argument ([here](https://github.com/status-im/nwaku/blob/master/waku/v2/node/peer_manager/peer_manager.nim#L269)):

```nim
proc dialPeer*(pm: PeerManager, remotePeerInfo: RemotePeerInfo, proto: string, dialTimeout = defaultDialTimeout): Future[Option[Connection]] {.async.} =
  # Dial a given peer and add it to the list of known peers
  # @TODO check peer validity and score before continuing. Limit number of peers to be managed.
  
  # First add dialed peer info to peer store, if it does not exist yet...
  if not pm.hasPeer(remotePeerInfo.peerId, proto):
    trace "Adding newly dialed peer to manager", peerId = remotePeerInfo.peerId, addr = remotePeerInfo.addrs[0], proto = proto
    pm.addPeer(remotePeerInfo, proto)
  
  if remotePeerInfo.peerId == pm.switch.peerInfo.peerId:
    # Do not attempt to dial self
    return none(Connection)

  return await pm.dialPeer(remotePeerInfo.peerId, remotePeerInfo.addrs, proto, dialTimeout)
```

* With `PeerID` as an argument ([here](https://github.com/status-im/nwaku/blob/master/waku/v2/node/peer_manager/peer_manager.nim#L284)):

```nim
proc dialPeer*(pm: PeerManager, peerId: PeerID, proto: string, dialTimeout = defaultDialTimeout): Future[Option[Connection]] {.async.} =
  # Dial an existing peer by looking up it's existing addrs in the switch's peerStore
  # @TODO check peer validity and score before continuing. Limit number of peers to be managed.
   
  if peerId == pm.switch.peerInfo.peerId:
    # Do not attempt to dial self
    return none(Connection)

  let addrs = pm.switch.peerStore[AddressBook][peerId]
  if addrs.len == 0:
    return none(Connection)

  return await pm.dialPeer(peerId, addrs, proto, dialTimeout)
```

If you inspect the content of the methods... both methods do two different things depending on the type of the peer argument 🎉

That "method overload trap" is exactly the issue that drove me crazy during the refactoring; and why I have been saying that we need to do a review/refactoring/cleanup in the PeerManager code.

Reply:
DISCLAIMER: I did not go into the rabbit hole to see why the `PeerId` method does not work as the other method.